### PR TITLE
read instead of browse (3 to 5x faster here)

### PR DESCRIPTION
**Pull request by ccomb**
OCA#216 for v8

### CODE REVIEW

**Comment by gdgellatly** [comment]
But what if there is more than one id?

I'm not keen on a function signature that indicates it takes multiple ids,
only returning the first result even if today it only ever receives a list
of one.

If performance really mattered then this is the kind of place that the
tradition has been to use direct SQL statements.

**Comment by ccomb** [comment]
Thanks for your comment, I also hate `[0]` in openerp code. It's most of the time a source of bugs. However here there is no chance the function receive many ids because it's a stored field trigger that are treated one by one. I can improve a little bit anyway by merging the different possible results.

**Comment by pedrobaeza** [comment]
In a Fabien's statement, he said that browse it at least the same fast as read. Why is this difference?

**Comment by gdgellatly** [comment]
because .id must access two table and two sets of access rules.  Then there
is the setup of the for loop (which can be overcome) To me it is an
optimisation of browse issue.  If .id is requested, don't read another
table, just take it from the existing one.

File: addons/sale/sale.py [hunk 1]
```diff
@@ -175,10 +175,9 @@ def _invoiced_search(self, cursor, user, obj, name, args, context=None):
         return [('id', 'in', [x[0] for x in res])]
 
     def _get_order(self, cr, uid, ids, context=None):
-        result = {}
-        for line in self.pool.get('sale.order.line').browse(cr, uid, ids, context=context):
-            result[line.order_id.id] = True
-        return result.keys()
+        lines = self.pool.get('sale.order.line')
```
Comment (pedrobaeza) [inline addons/sale/sale.py]: Use `line_obj` or `line_model` to name the variable for following conventions.

Comment (ccomb) [inline addons/sale/sale.py]: ok, just doing

**Comment by ccomb** [comment]
I'm sorry to say that I don't trust Fabien's statements: I only trust what I can measure.
It's easy to reproduce: add a timing code around the function, then create a sale order. Then modify a sale order line (change the qty).

On v7 without the patch, the function is run twice:
- 1.98ms
- 1.10ms

With the patch:
- 0.44ms
- 0.45ms

It seems small, but when you run it thousands of times in unittest or in cron operations, it's a real benefit. For a customer I'm also adding caching to this function so that it is only run once. But it needs a different implementation of ormcache (the one found in https://bitbucket.org/anybox/anybox_perf/)

**Comment by ccomb** [comment]
On v8 the difference is:

without the patch:
- 4.18ms
- 0.11ms

With the patch:
- 1.16ms
- 1.18ms

This is still 2x faster with the patch, however it shows there is a better caching in v8.

**Comment by bwrsandman** [comment]
You might consider using an OrderedDict since reads are supposed to respect an order. `list(set())` is good for removing duplicates, but the order becomes non-deterministic for `len(set(results)) > 1`

**Comment by ccomb** [comment]
@bwrsandman this is not a read, this is a trigger for a stored function field, which is supposed to return a list of ids to update. The original implementation is based on a dict keys so this not ordered either.
Anyway this function eventually will always receives only 1 id, and will always returns 1 id...

**Comment by bwrsandman** [comment]
Got it.

I approve the use of read instead of browse for cases which need optimization.
:+1:

**Comment by pedrobaeza** [comment]
:+1:

**Comment by gurneyalex** [comment]
:+1:

**Comment by gdgellatly** [comment]
Just a question. I am still not sure about this single I'd gurentee.  I
thought if defer computation was set. Like for imports or when field
definitions or defaults change it could receive up to 40 in v7 and loads
more in v8.  But I never really tested it, just something I recall from a
day bug session in an upgrade.

**Comment by ccomb** [comment]
You're right, I've found the place were the function is called [for v7](https://github.com/OCA/OCB/blob/7.0/openerp/osv/orm.py#L4667) and [for v8](https://github.com/OCA/OCB/blob/8.0/openerp/models.py#L4314).
It clearly shows that the ids directly comes from the ids of the `write()` function where the `_store_get_values` is called. It's difficult to obtain multiple ids through the web interface because it would require to change many records at once, but it is be possible through the API.
(So my first implementation was indeed broken.)

**Comment by pedrobaeza** [comment]
But as far as I have seen, you have fixed it, right?

**Comment by ccomb** [comment]
Yes I did, it now loops on the given order line ids and returns the list of corresponding (and non duplicated)  order_id using a generator expression inside a set.

**Comment by pedrobaeza** [comment]
I proceed with the merge then.

Thanks.

**Comment by ccomb** [comment]
thx all for the review!

**Pull request merged**